### PR TITLE
Typescript-strict-plugin

### PR DESCRIPTION
**Pull request by MGibson1**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-6684

## 📔 Objective

Use [typescript-strict-plugin](https://github.com/allegro/typescript-strict-plugin) to allow iterative approach to enabling strict everywhere.

This PR obviously touches a huge number of files, but the commits are broken down such that the last can effectively be ignored. It's the result of runnning the script that excludes offending files initially.

Benefits of this approach is that new files will automatically be required to be strict compliant. This is enforced by both a required pipeline job and updated tsconfig and vscode settings to use tsc-strict in IDE.

Be sure to read up on the history of this change and the [guidelines](https://contributing.bitwarden.com/architecture/adr/typescript-strict/#guidelines) for how to migrate existing code available in [ADR 014](https://contributing.bitwarden.com/architecture/adr/typescript-strict/)

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: apps/cli/src/service-container/tsconfig.json
Comment (MGibson1) [inline apps/cli/src/service-container/tsconfig.json:1]: This tsconfig turned on a subset of strict checks, but the files controlled by it are not yet compliant, so unfortunately `service-container.ts` needed an ignore comment.

File: scripts/test-types.js [hunk 1]
```diff
@@ -21,9 +21,14 @@ const files = getFiles(path.join(__dirname, "..", "libs")).filter((file) => {
...
 });
 
-concurrently(
-  files.map((file) => ({
+concurrently([
+  {
+    // run the strict type check plugin until we're fully converted, then update tsconfig.json to use strict
+    name: "typescript-strict-plugin",
+    command: "npx tsc-strict",
+  },
```
Comment (MGibson1) [inline scripts/test-types.js:29]: enforces adherence to strict rules while honoring the `@ts-strict-ignore` comment.

File: apps/browser/tsconfig.json
Comment (Hinton) [inline apps/browser/tsconfig.json:1]: I don't believe we need to override plugins in every tsconfig file? It seems to work fine with just the plugin in the root.

**Review by Hinton — COMMENTED** [review]
You should link the ADR where we discussed the different approaches.

I'm looking at the first 6 commits and testing the functionality. It seems to behave as expected without vscode workspaces, but not for workspaces. You need to update the lib for the workspace file too. (Open the workspace, select typescript version ....)

File: apps/browser/tsconfig.json
Comment (MGibson1) [reply apps/browser/tsconfig.json:1]: In my testing I did. vscode was honoring the closest tsconfig file when evaluating a given ts file

Comment (Hinton) [reply apps/browser/tsconfig.json:1]: Try removing it from say apps/web and restarting typescript service. You also have tsconfig files in libs/common and it's not needed there?

Comment (MGibson1) [reply apps/browser/tsconfig.json:1]: libs generally extend a shared tsconfig, which is updated here. It turns out that web is the same way -- extending another config. I didn't realize that so I was able to remove a few more tsconfigs from being updated. The ones that remain are standalone.

File: scripts/test-types.js (hunk 1 above)
Comment (Hinton) [reply scripts/test-types.js:29]: This would only run for libs.

**Review by Hinton — APPROVED** [review]
https://github.com/bitwarden/clients/pull/12235/files/67f67afc6a5ada125fd9536307809b081afbb0b4 looks good.

This will need championing to get done, a good starting point is probably to revive my old PR, https://github.com/bitwarden/clients/pull/8092 since it resolves most abstract classes.

Reviewers only need to review their teams owned files for the last two commits.

**Review by mzieniukbw — APPROVED** [review]

**Review by willmartian — APPROVED** [review]

**Review by dani-garcia — DISMISSED** [review]

**Review by rr-bw — APPROVED** [review]

**Review by BTreston — APPROVED** [review]

**Review by djsmith85 — DISMISSED** [review]

**Review by jprusik — DISMISSED** [review]

**Review by shane-melton — APPROVED** [review]

**Review by willmartian — APPROVED** [review]
@MGibson1 There are a few failing tests; I think they are new files without the ts strict comment. That said, I am fine with merging this and doing a followup PR to fix such cases so this doesn't get stuck in review forever.

**Review by jprusik — APPROVED** [review]

**Comment by MGibson1** [comment]
you guys are too fast, we need to fix the files since it's a required pipeline job. I think we'll get there, though 😄

**Review by shane-melton — APPROVED** [review]

**Review by rr-bw — APPROVED** [review]

**Review by BTreston — APPROVED** [review]

**Review by mzieniukbw — APPROVED** [review]

**Review by willmartian — APPROVED** [review]

**Review by dani-garcia — APPROVED** [review]

**Review by djsmith85 — APPROVED** [review]

**Pull request merged**